include/cuco/detail/open_addressing/open_addressing_impl.cuh

```diff
@@ -681,6 +681,73 @@ class open_addressing_impl {
     return output_begin + h_num_out;
   }
 
+  /**
+   * @brief Executes a callback on every filled element in the container.
+   *
+   * @note Passes an un-incrementable input iterator to the element whose key is filled
+   * to the callback.
+   *
+   * @tparam CallbackOp Unary callback functor or device lambda
+   *
+   * @param callback_op Function to call on every filled element in the container
+   * @param stream CUDA stream used for this operation
+   */
+  template <typename CallbackOp>
+  void for_each_async(CallbackOp&& callback_op, cuda::stream_ref stream)
+  {
+    using const_iterator = typename storage_ref_type::const_iterator;
+
+    auto const is_filled = open_addressing_ns::detail::slot_is_filled<has_payload, key_type>{
+      this->empty_key_sentinel(), this->erased_key_sentinel()};
+
+    thrust::for_each(
```

Comment:
Ah, good use of `thrust::for_each` here. IIRC this function as is is still synchronous w.r.t. the host. You'd have to use `thrust::async::for_each` and/or `thrust::cuda::par_nosync` execution policy.

Reply:
Thanks! Resolved in https://github.com/NVIDIA/cuCollections/pull/565/commits/ded6180e14564ff1e88a2a9022d5e0dbda85b7b8.

Reply:
Let's switch to `cub::DeviceFor`. `thrust::for_each` will not use the map allocator to deal with temporary memory allocation/deallocation.

Reply:
Does `for_each` need any temp allocations?

Reply:
TBH, I don't know. If the underlying algo is [cub::DeviceFor](https://nvidia.github.io/cccl/cub/api/structcub_1_1DeviceFor.html#_CPPv4N3cub9DeviceForE), then probably yes. If not, it might be switched to `cub::DeviceFor` in the future and the only way to make sure that we are using the given allocator is to invoke APIs with explicit memory control. 

With that, cuco details probably always want to use CUB instead of thrust.